### PR TITLE
Initial script edit

### PR DESCRIPTION
**Pull request by caelanradford**
This adds the initial edited script from CodonTilingPrimers and an updated README to describe what the changes do. Here is a short summary: 
- You must now pass a csv file as an argument that specifies what mutants to make at what sites 
- You must now pass a codon frequency csv file as an argument that determines what codon will to mutated to for each amino acid 
- The gene sequence given in a text file should now be uppercase for the WHOLE gene, including the start and stop codons 
- The gene sequence given in a text file no longer necessarily has to have lower case sequence before and after the gene; there must only be enough room for primer overhangs before the first codon mutagenized and after the last codon mutagenized
- As is written right now, there is no option for degenerate codons, which is why this is a new repository. This functionality could be added back later 

More functionality could be added later, like randomly picking codons for an amino acid as an option, or picking weighted on frequencies, etc. If this edit looks ok, I'll make a test case with BF520 primers and add them to the repo. I am still also making sure it is working properly; just wanted to make sure this approach looked ok in the meantime.

### CODE REVIEW

**Comment by jbloom** [comment]
@khdusenbury or @fwelsh, are one or both of you able to review this?

**Comment by khdcrawford** [comment]
I am happy to review, but probably tomorrow (or tonight at the earliest).

**Comment by fwelsh** [comment]
I can look over it this afternoon as well

File: README.md
Comment (khdcrawford) [inline README.md]: I _think_ the CodonTilingPrimers script can mutate all codons to any specified other codon? That's not necessary to mention, but I would mention that it's using degenerate nucleotides as a contrast to the new script, so maybe say "This protocol randomly mutates each codon to all other possible amino acids using degenerate nucleotides such as `NNS`" 

Because also with `NNS`, we're not mutating to all other codons, but to all other amino acids.

File: README.md
Comment (khdcrawford) [inline README.md]: Probably say "all possible amino acid mutations" since we aren't necessarily doing al codons anymore.

File: README.md
Comment (khdcrawford) [inline README.md]: If a site has multiple mutants, does it need to be listed multiple times, or can you supply a list to the 'mutant' column?

File: README.md
Comment (khdcrawford) [inline README.md:25]: Haven't look at code yet, so not sure if it's worth changing now or how it would work, but I'm pretty sure you can just import codon frequency tables (cai tables I think they're called, but I'm not sure what that stands for) using BioPython.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Could you add new lines to make this doc string easier to read? In python, lines should technically be no more than 79 characters long.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Also please continue these strings onto new lines to maintain line length <= 79 characters.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Is there a reason my `S`, `K`, and `M` are the only ambiguous nucleotides included?

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py:175]: I think this should be using `nt.isupper()`. Technically `istitle()` works, but you're just comparing one nt at a time, so `isupper()` seems better.

You can also probably use `isupper()` on the seq somehow and not have to iterate to get all upper case letters back out, but I haven't confirmed the syntax for that.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: This is a small thing, but I think @jbloom told me recently that `assert` statements are good for debugging, but for actual error checking that you want to be part of your final code, using `raise ValueError` is better. @jbloom can clarify perhaps?

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Do you use `n` anywhere?

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: This is a small and potentially just a preference thing, but if you're going to have a string with both ' and ", I think it's best to have the " on the outside and the ' specifying the inside quotes. Like: "`mutations_csv` must have column 'site'"

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: You already set this variable in line 93. Just set it in one place.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Oh here is where you use `n`. Maybe just define `n` closer to where you use it.

Comment (khdcrawford) [inline create_primers.py]: This also took me a long time to parse, maybe provide a comment above it saying what each part is measuring.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: I'm not sure, but I think this is maybe where you could use the built in BioPython cai tables?

I _think_ that functionality might allow you to go from aa -> codon and back more easily.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py:220]: Probably not necessary now, but in terms of extensions of this code, you could imagine adding a `--codon_choice` flag and have it be either `max` or `random` to specify if the codon chosen is always the one with the max frequency or if you choose randomly (maybe for random choice, you'd want to set a threshold minimum frequency needed for a codon to be included in the `back_t_dict`)

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Do you do anything with this list?

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Please break onto multiple lines!

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: You use this in a lot of places, but I'm just commenting here. The "%s", "%d", etc. string formatting is a hold over from the earlier script, but it is also very Python 2. 

I'd recommend switching all your strings to f string formatting, which is much more readable in my opinion too. So the code would be:

`primer = f"{seq[i - initial_flanklength : i]}{codon_insert}{seq[i + 3 : i + 3 + initial_flanklength]}"`

Like in the line below.

I know there are lots of places where this change would need to be made, but I think it's worth making in order to be consistent and up-to-date.

**Review by khdcrawford — CHANGES_REQUESTED** [review]
Overall, this looks good Caelan. I there are a couple things to clean up, but most of it is just formatting. I would definitely add some tests and I do think the functionality of picking max vs. random codon would be good, probably not necessary, so I'd say only do it if you want.

File: README.md
Comment (caelanradford) [reply README.md]: I chose to have each mutation be a line in the file so that it would be possible later to add the ability to specify how many codons to generate for that mutation or how to choose the codons (which would take adding another column).

Comment (khdcrawford) [reply README.md]: Sounds good, you should just make it clear then that if a site has multiple mutants, it needs to be listed separately for each mutant. That's not clear from the description.

File: README.md
Comment (caelanradford) [reply README.md:25]: I think BioPython can be used to generate codon frequency tables or use them to score sequences on their codon usage, but I did not find any well documented functionality to back translate using them. If you can find something like that I would use it.

File: create_primers.py
Comment (caelanradford) [reply create_primers.py]: This was a copy over from the old script. We could add more, which ones do you want to add?

File: README.md
Comment (khdcrawford) [reply README.md:25]: I need to think more about the back translation, but why exactly is it necessary? Aren't you specifying mutants by aa, so why do you need to back translate from codon?

File: create_primers.py
Comment (khdcrawford) [reply create_primers.py]: I guess my main point is that this should be deliberate and not just a copy/paste, so it should include what you think is necessary for the script (or maybe all of the degenerate nucleotides if you think at one point this script might allow degenerate nucleotide functionality).

File: README.md
Comment (caelanradford) [reply README.md:25]: If you specify what mutants you want by aa, then unless you want to make primers for each codon for that aa, you have to pick a codon (back translate the aa to a codon). The codon frequency table is used so that the codon picked isn't just random, but is the most frequent codon for that aa in whatever organism you supply the table for. I am planning on adding an option for just random, or semi random like you suggested.

Comment (khdcrawford) [reply README.md:25]: Gotcha. I thought that we had used a built in function for this in `serologyseq` (https://github.com/jbloomlab/serologyseq), but it looks like we did write our own dictionary to map amino acids and codons. We did just use built in cai tables though, so there might be something helpful in the `design_oligos.py` script (https://github.com/jbloomlab/serologyseq/blob/master/src/design_oligos.py) if you didn't want to have to have the user input their own codon_frequency table, but it's not as easy as I thought it was, so what you have with the user inputting their own frequency table is also fine.

File: create_primers.py
Comment (caelanradford) [reply create_primers.py]: It was from an earlier version that I forgot to delete, thanks.

File: README.md
Comment (caelanradford) [reply README.md]: I tried to clarify this more

**Comment by caelanradford** [comment]
I made a lot of Kate's suggested edits. I got rid of the old string syntax, and changed them to f strings. I made the arguments comments more readable. I added a few lines of code in different places to break things up for the sake of readability. Also slightly changed some wording in the README. I also fixed a bug that I had missed before that was making the first codon primers were generated before too late in the gene. 

I am planning on adding some of the options Kate mentioned. I am planning on adding the ability to specify whether you want codons chosen based on frequencies or chosen at random, or some combination option (random pick from frequencies above some threshold). I am also planning on adding an option to choose how many codons you want for each mutation. One of the reasons I chose to have the mutations passed in a csv with a row for each mutation rather than each site is so that you could potentially change the number of codons you are targeting by mutation rather than just by site.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Sorry for more knit-picky style edits, but these continuation lines should be indented either two tabs or equal with the original open parenthesis. 

I suggest running `flake8` on your code as it will catch these style things and also give somewhat helpful error messages as to how to fix the style.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Same indent thing here regarding the continuation line (I think it's best to be indented equal to the open parenthesis, but I think 2 tabs is technically also acceptable, but I don't quite remember).

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: I'm not sure if this will cause issues or not, but you're now no longer rounding the Tm and it's not a string anymore. 

I haven't dug into the code enough to know if this is an issue, but worth keeping in mind/maybe writing some tests to confirm that's okay?

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py:234]: TmNN is now no longer a string, so I'm not sure if it's necessary to have `float()` around it, but I'd check all of this and make sure it doesn't mess up things anywhere that you're not rounding `TmNN` anymore. I could see some benefit to strictly defining how many digits `TmNN` includes, which you're not doing anymore.

File: create_primers.py
Comment (khdcrawford) [inline create_primers.py]: Do you need the plate functionality? Or maybe it should be an argument for "order in plate" or not? I'm not sure how you order these things and if just a long list is sufficient, but I assume Adam had this functionality for a reason? But that may also be out-of-date.

**Review by khdcrawford — APPROVED** [review]
Looks great @caelanradford. I had a couple more syntax things, and then I think you should run `flake8` on this and make sure it passes before you merge, but I'm going to approve so you can merge when you're ready.

File: create_primers.py
Comment (caelanradford) [reply create_primers.py]: I checked when playing around with the code, and it doesn't seem to do anything when no longer rounding it. It does make it easier to look at primer temps if you decide to print them out to check they are coming out in the range you want, so maybe I'll revert the change.

File: create_primers.py
Comment (caelanradford) [reply create_primers.py]: The in-plate ordering functionality was for ordering from IDT in plates and then hand mixing the primers. It seems the best options now are to order oligo pools pre-pooled, which we just need a list of the names and primers for.

Comment (khdcrawford) [reply create_primers.py]: Oh yeah, that sounds way better! Sweet.

**Pull request merged**
